src/main/java/com/code_intelligence/jazzer/mutation/mutator/proto/BuilderAdapters.java

```diff
@@ -16,64 +16,130 @@
 
 package com.code_intelligence.jazzer.mutation.mutator.proto;
 
-import static java.lang.String.format;
+import static java.util.Collections.singletonList;
 
 import com.google.protobuf.Descriptors.FieldDescriptor;
 import com.google.protobuf.MapEntry;
 import com.google.protobuf.Message.Builder;
 import java.util.AbstractList;
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Map.Entry;
 
 final class BuilderAdapters {
+  private BuilderAdapters() {}
+
   static <T extends Builder, U> List<U> makeMutableRepeatedFieldView(
       T builder, FieldDescriptor field) {
     return new AbstractList<U>() {
+      // O(1)
       @Override
       public U get(int index) {
         return (U) builder.getRepeatedField(field, index);
       }
 
+      // O(1)
       @Override
       public int size() {
         return builder.getRepeatedFieldCount(field);
       }
 
+      // O(1)
+      @Override
+      public boolean add(U element) {
+        builder.addRepeatedField(field, element);
+        return true;
+      }
+
+      // O(1)
       @Override
-      public boolean add(U u) {
-        builder.addRepeatedField(field, u);
+      public void add(int index, U element) {
+        addAll(index, singletonList(element));
+      }
+
+      // O(size() + other.size())
+      public boolean addAll(int index, Collection<? extends U> other) {
+        // This was benchmarked against the following implementation and found to be faster in all
+        // cases (up to 4x on lists of size 1000):
+        //
+        // for (U element : other) {
+        //   builder.addRepeatedField(field, element);
+        // }
+        // Collections.rotate(subList(index, size()), other.size());
```

Comment:
I like the perf annotations a lot! That should make things easier when trying to improve the mutator framework in the long run :).